### PR TITLE
Refactor pytest fixtures for ray core

### PR DESCRIPTION
**Pull request by guoyuhong**
## What do these changes do?
Move fixtures to a file fixture.py and use `pytest.mark.parametrize` as much as possible. We have following benefits:
1. less duplicated code.
2. ability to specify some ray start arguments to most test cases. 

## Related issue number
N/A

### CODE REVIEW

File: python/ray/tests/test_basic.py [hunk 1]
```diff
@@ -28,28 +28,18 @@
 import ray
 import ray.tests.cluster_utils
 import ray.tests.utils
+from ray.tests.fixtures import (
+    ray_start_2_cpus,
+    ray_start_cluster,
+    ray_start_regular,
+    shutdown_only,
+)  # noqa: F401,F811
```
Comment (robertnishihara) [inline python/ray/tests/test_basic.py]: What's the linting error here? We shouldn't need the `noqa`. Same comment applies in other places.

File: python/ray/tests/fixtures.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    address_info = ray_start_with_parameter(request)
+    yield address_info
+    # The code after the yield will run as teardown code.
+    ray.shutdown()
+
+
+@pytest.fixture
+def ray_start_2_cpus(request):
+    parameter = getattr(request, "param", {})
+    parameter['num_cpus'] = 2
```
Comment (robertnishihara) [inline python/ray/tests/fixtures.py]:
```suggestion
    parameter["num_cpus"] = 2
```

and in other places

**Review by robertnishihara — COMMENTED** [review]
Nice! This will be much cleaner. I haven't looked carefully over the whole thing. Are there any behavior changes here?

File: python/ray/tests/test_basic.py (hunk 1 above)
Comment (guoyuhong) [reply python/ray/tests/test_basic.py]: The lint error looks strange to me. May the usage is not correct here.
```
./python/ray/tests/test_basic.py:31:1: F401 'ray.tests.fixtures.ray_start_2_cpus' imported but unused
./python/ray/tests/test_basic.py:31:1: F401 'ray.tests.fixtures.ray_start_cluster' imported but unused
./python/ray/tests/test_basic.py:31:1: F401 'ray.tests.fixtures.ray_start_regular' imported but unused
./python/ray/tests/test_basic.py:31:1: F401 'ray.tests.fixtures.shutdown_only' imported but unused
./python/ray/tests/test_basic.py:42:31: F811 redefinition of unused 'ray_start_regular' from line 31
./python/ray/tests/test_basic.py:109:32: F811 redefinition of unused 'ray_start_regular' from line 31
./python/ray/tests/test_basic.py:312:32: F811 redefinition of unused 'ray_start_regular' from line 31
./python/ray/tests/test_basic.py:340:52: F811 redefinition of unused 'ray_start_regular' from line 31
./python/ray/tests/test_basic.py:372:52: F811 redefinition of unused 'ray_start_regular' from line 31
./python/ray/tests/test_basic.py:387:18: F811 redefinition of unused 'shutdown_only' from line 31
./python/ray/tests/test_basic.py:415:29: F811 redefinition of unused 'ray_start_regular' from line 31
./python/ray/tests/test_basic.py:445:39: F811 redefinition of unused 'ray_start_regular' from line 31
./python/ray/tests/test_basic.py:462:25: F811 redefinition of unused 'ray_start_2_cpus' from line 31
./python/ray/tests/test_basic.py:605:23: F811 redefinition of unused 'ray_start_regular' from line 31
./python/ray/tests/test_basic.py:690:34: F811 redefinition of unused 'shutdown_only' from line 31
./python/ray/tests/test_basic.py:747:36: F811 redefinition of unused 'shutdown_only' from line 31
./python/ray/tests/test_basic.py:815:21: F811 redefinition of unused 'shutdown_only' from line 31
./python/ray/tests/test_basic.py:873:23: F811 redefinition of unused 'ray_start_regular' from line 31
./python/ray/tests/test_basic.py:884:36: F811 redefinition of unused 'ray_start_regular' from line 31
./python/ray/tests/test_basic.py:894:19: F811 redefinition of unused 'ray_start_regular' from line 31
./python/ray/tests/test_basic.py:903:15: F811 redefinition of unused 'ray_start_regular' from line 31
./python/ray/tests/test_basic.py:958:25: F811 redefinition of unused 'ray_start_regular' from line 31
./python/ray/tests/test_basic.py:979:34: F811 redefinition of unused 'shutdown_only' from line 31
./python/ray/tests/test_basic.py:1010:35: F811 redefinition of unused 'ray_start_regular' from line 31
./python/ray/tests/test_basic.py:1053:42: F811 redefinition of unused 'ray_start_regular' from line 31
./python/ray/tests/test_basic.py:1080:24: F811 redefinition of unused 'ray_start_2_cpus' from line 31
./python/ray/tests/test_basic.py:1124:23: F811 redefinition of unused 'ray_start_cluster' from line 31
./python/ray/tests/test_basic.py:1150:31: F811 redefinition of unused 'ray_start_cluster' from line 31
./python/ray/tests/test_basic.py:1191:35: F811 redefinition of unused 'ray_start_regular' from line 31
./python/ray/tests/test_basic.py:1257:28: F811 redefinition of unused 'ray_start_regular' from line 31
./python/ray/tests/test_basic.py:1272:25: F811 redefinition of unused 'ray_start_2_cpus' from line 31
./python/ray/tests/test_basic.py:1409:34: F811 redefinition of unused 'ray_start_cluster' from line 31
./python/ray/tests/test_basic.py:1496:21: F811 redefinition of unused 'shutdown_only' from line 31
./python/ray/tests/test_basic.py:1579:31: F811 redefinition of unused 'shutdown_only' from line 31
./python/ray/tests/test_basic.py:1655:37: F811 redefinition of unused 'shutdown_only' from line 31
./python/ray/tests/test_basic.py:1708:18: F811 redefinition of unused 'shutdown_only' from line 31
./python/ray/tests/test_basic.py:1861:20: F811 redefinition of unused 'shutdown_only' from line 31
./python/ray/tests/test_basic.py:1872:26: F811 redefinition of unused 'ray_start_cluster' from line 31
./python/ray/tests/test_basic.py:1890:31: F811 redefinition of unused 'shutdown_only' from line 31
./python/ray/tests/test_basic.py:1941:36: F811 redefinition of unused 'ray_start_cluster' from line 31
./python/ray/tests/test_basic.py:2067:27: F811 redefinition of unused 'ray_start_cluster' from line 31
./python/ray/tests/test_basic.py:2103:31: F811 redefinition of unused 'ray_start_cluster' from line 31
./python/ray/tests/test_basic.py:2159:32: F811 redefinition of unused 'shutdown_only' from line 31
./python/ray/tests/test_basic.py:2234:25: F811 redefinition of unused 'ray_start_regular' from line 31
./python/ray/tests/test_basic.py:2269:25: F811 redefinition of unused 'ray_start_regular' from line 31
./python/ray/tests/test_basic.py:2307:25: F811 redefinition of unused 'ray_start_cluster' from line 31
./python/ray/tests/test_basic.py:2326:43: F811 redefinition of unused 'ray_start_cluster' from line 31
./python/ray/tests/test_basic.py:2370:27: F811 redefinition of unused 'shutdown_only' from line 31
./python/ray/tests/test_basic.py:2508:28: F811 redefinition of unused 'shutdown_only' from line 31
./python/ray/tests/test_basic.py:2531:32: F811 redefinition of unused 'shutdown_only' from line 31
./python/ray/tests/test_basic.py:2556:18: F811 redefinition of unused 'shutdown_only' from line 31
./python/ray/tests/test_basic.py:2633:30: F811 redefinition of unused 'shutdown_only' from line 31
./python/ray/tests/test_basic.py:2649:27: F811 redefinition of unused 'ray_start_2_cpus' from line 31
./python/ray/tests/test_basic.py:2668:35: F811 redefinition of unused 'shutdown_only' from line 31
./python/ray/tests/test_basic.py:2694:20: F811 redefinition of unused 'ray_start_2_cpus' from line 31
./python/ray/tests/test_basic.py:2742:34: F811 redefinition of unused 'shutdown_only' from line 31
./python/ray/tests/test_basic.py:2750:46: F811 redefinition of unused 'ray_start_regular' from line 31
./python/ray/tests/test_basic.py:2771:28: F811 redefinition of unused 'ray_start_regular' from line 31
./python/ray/tests/test_basic.py:2819:31: F811 redefinition of unused 'shutdown_only' from line 31
```

Comment (guoyuhong) [reply python/ray/tests/test_basic.py]: I checked pytest document. I need to move the fixtures from `fixtures.py` to `conftest.py`. Then we can use all the fixtures without imports.

**Comment by guoyuhong** [comment]
Set `object_store_memory=10**8` according to https://github.com/ray-project/ray/pull/4385#issuecomment-474336671 .

**Comment by guoyuhong** [comment]
Ready for review.

File: python/ray/tests/conftest.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+@pytest.fixture
+def ray_start_cluster_2_nodes(request):
+    param = getattr(request, "param", {})
+    with _ray_start_cluster(do_init=True, num_nodes=2, **param) as res:
+        yield res
+
+
+@pytest.fixture
+def ray_start_object_store_memory(request):
```
Comment (raulchen) [inline python/ray/tests/conftest.py]: This function can also be based on `_ray_start`

**Review by raulchen — APPROVED** [review]
I left a small comment, otherwise LGTM

File: python/ray/tests/conftest.py (hunk 3 above)
Comment (raulchen) [reply python/ray/tests/conftest.py]: nvm, didn't notice that this function yield the store size. Then it's fine.

**Pull request merged**